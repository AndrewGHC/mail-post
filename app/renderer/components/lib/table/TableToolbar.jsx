import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import Tooltip from '@material-ui/core/Tooltip';
import DeleteIcon from '@material-ui/icons/Delete';
import { lighten } from '@material-ui/core/styles/colorManipulator';

const toolbarStyles = theme => ({
  root: {
    paddingRight: theme.spacing.unit,
  },
  highlight: {
    color: theme.palette.secondary.main,
    backgroundColor: lighten(theme.palette.secondary.light, 0.85),
  },
  spacer: {
    flex: '1 1 100%',
  },
  actions: {
    color: theme.palette.text.secondary,
  },
  title: {
    flex: '0 0 auto',
  },
});

const EnhancedTableToolbar = (props) => {
  const {
    numSelected,
    classes,
    onClickDelete,
    title,
  } = props;

  return (
    <Toolbar
      className={classNames(classes.root, {
        [classes.highlight]: numSelected > 0,
      })}
    >
      <div className={classes.title}>
        {numSelected > 0 ? (
          <Typography color="inherit" variant="subtitle1">
            {numSelected}
            {' '}
selected
          </Typography>
        ) : (
          <Typography variant="h6">{title}</Typography>
        )}
      </div>
      <div className={classes.spacer} />
      <div className={classes.actions}>
        {numSelected > 0 && (
          <Tooltip title="Delete">
            <IconButton aria-label="Delete" onClick={onClickDelete}>
              <DeleteIcon />
            </IconButton>
          </Tooltip>
        )}
      </div>
    </Toolbar>
  );
};

EnhancedTableToolbar.propTypes = {
  classes: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  numSelected: PropTypes.number.isRequired,
  onClickDelete: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
};

const StyledEnhancedTableToolbar = withStyles(toolbarStyles)(EnhancedTableToolbar);

export default StyledEnhancedTableToolbar;
