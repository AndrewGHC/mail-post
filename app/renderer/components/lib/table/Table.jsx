import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Table, {
  TableBody,
  TableCell,
  TablePagination,
  TableRow,
} from 'material-ui/Table';
import Paper from 'material-ui/Paper';
import Checkbox from 'material-ui/Checkbox';
import moment from 'moment';
import { clone } from 'ramda';

import sort from './sort-data';
import { generateDeleteLists } from '../../../graphql';
import TableHeader from './TableHeader';
import TableToolbar from './TableToolbar';

class EnhancedTable extends Component {
  constructor(props) {
    super(props);

    this.state = {
      order: 'desc',
      orderBy: 'createdAt',
      selected: [],
      data: props.data && props.data.lists ? props.data.lists : [],
      page: 0,
      rowsPerPage: 5,
    };

    this.handleRequestSort = this.handleRequestSort.bind(this);
    this.handleSelectAllClick = this.handleSelectAllClick.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleChangePage = this.handleChangePage.bind(this);
    this.handleChangeRowsPerPage = this.handleChangeRowsPerPage.bind(this);
    this.isSelected = this.isSelected.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.data !== nextProps.data) {
      const lists = clone(nextProps.data.lists).map(({ createdAt, ...rest }) => ({
        ...rest,
        createdAt: new Date(createdAt).getTime(),
      }));
      const { data } = sort(lists, this.state.order, this.state.orderBy);
      this.setState({ data });
    }
  }

  handleRequestSort(event, property) {
    const { data, order } = sort(this.state.data, this.state.order, this.state.orderBy, property);
    this.setState({ data, order, orderBy: property });
  }

  handleSelectAllClick(event, checked) {
    if (checked) {
      this.setState({ selected: this.state.data.map(n => n.id) });
      return;
    }
    this.setState({ selected: [] });
  }

  handleClick(event, id) {
    const { selected } = this.state;
    const selectedIndex = selected.indexOf(id);
    let newSelected = [];

    if (selectedIndex === -1) {
      newSelected = newSelected.concat(selected, id);
    } else if (selectedIndex === 0) {
      newSelected = newSelected.concat(selected.slice(1));
    } else if (selectedIndex === selected.length - 1) {
      newSelected = newSelected.concat(selected.slice(0, -1));
    } else if (selectedIndex > 0) {
      newSelected = newSelected.concat(
        selected.slice(0, selectedIndex),
        selected.slice(selectedIndex + 1),
      );
    }

    this.setState({ selected: newSelected });
  }

  handleChangePage(event, page) {
    this.setState({ page });
  }

  handleChangeRowsPerPage(event) {
    this.setState({ rowsPerPage: event.target.value });
  }

  isSelected(id) {
    return this.state.selected.indexOf(id) !== -1;
  }

  handleDelete() {
    const { selected } = this.state;
    this.props.MUTATION_DELETE_LISTS(generateDeleteLists(selected));
    this.setState({ selected: [] });
  }

  render() {
    const {
      data,
      order,
      orderBy,
      selected,
      rowsPerPage,
      page,
    } = this.state;
    return (
      <Paper>
        <TableToolbar
          numSelected={selected.length}
          onClickDelete={this.handleDelete}
        />
        <div>
          <Table>
            <TableHeader
              numSelected={selected.length}
              order={order}
              orderBy={orderBy}
              onSelectAllClick={this.handleSelectAllClick}
              onRequestSort={this.handleRequestSort}
              rowCount={data.length}
            />
            <TableBody>
              {data.slice(page * rowsPerPage, (page * rowsPerPage) + rowsPerPage).map((n) => {
                const isSelected = this.isSelected(n.id);
                return (
                  <TableRow
                    hover
                    onClick={event => this.handleClick(event, n.id)}
                    role="checkbox"
                    aria-checked={isSelected}
                    tabIndex={-1}
                    key={n.id}
                    selected={isSelected}
                  >
                    <TableCell padding="checkbox">
                      <Checkbox checked={isSelected} />
                    </TableCell>
                    <TableCell padding="none">{n.name}</TableCell>
                    <TableCell numeric>{n.total_subscribers}</TableCell>
                    <TableCell numeric>{moment(n.createdAt).add(24, 'hours').format('LLL')}</TableCell>
                  </TableRow>
                );
              })}
            </TableBody>
          </Table>
        </div>
        <TablePagination
          component="div"
          count={data.length}
          rowsPerPage={rowsPerPage}
          page={page}
          backIconButtonProps={{
            'aria-label': 'Previous Page',
          }}
          nextIconButtonProps={{
            'aria-label': 'Next Page',
          }}
          onChangePage={this.handleChangePage}
          onChangeRowsPerPage={this.handleChangeRowsPerPage}
        />
      </Paper>
    );
  }
}

EnhancedTable.propTypes = {
  data: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  MUTATION_DELETE_LISTS: PropTypes.func.isRequired,
};

export default EnhancedTable;
