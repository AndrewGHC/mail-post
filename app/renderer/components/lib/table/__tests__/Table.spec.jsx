import React from 'react';
import td from 'testdouble';
import { shallow } from 'enzyme';
import TableRow from '@material-ui/core/TableRow';
import TablePagination from '@material-ui/core/TablePagination';

import Table from '../Table';
import TableToolbar from '../TableToolbar';

require('testdouble-jest')(td, jest);

describe('Table', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      data: [
        {
          id: '1',
          name: 'name',
          total_subscribers: '10',
          createdAt: '2015-03-25T12:00:00-06:30',
        },
        {
          id: '2',
          name: 'name',
          total_subscribers: '10',
          createdAt: '2015-03-25T12:00:00-06:30',
        },
      ],
      columnData: [
        {
          id: 'name',
          numeric: false,
          disablePadding: true,
          label: 'Name',
        },
        {
          id: 'total_subscribers',
          numeric: true,
          disablePadding: false,
          label: 'Subscribers',
        },
        {
          id: 'createdAt',
          numeric: true,
          disablePadding: false,
          label: 'Created',
        },
      ],
      deleteItem: td.function(),
      title: 'Test Table',
    };
    wrapper = shallow(<Table {...props} />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('when TableToolbar onClickDelete prop is called, calls prop deleteItem', () => {
    wrapper.find(TableToolbar).prop('onClickDelete')();
    td.verify(props.deleteItem(td.matchers.anything()));
  });

  it('when TableRow is selected, selected state contains item', () => {
    wrapper.find(TableRow).first().simulate('click');
    expect(wrapper.state('selected')).toEqual([props.data[0].id]);
  });

  it('when TableRow is unselected, selected state does not contain item', () => {
    wrapper.find(TableRow).first().simulate('click');
    wrapper.find(TableRow).first().simulate('click');
    expect(wrapper.state('selected')).toEqual([]);
  });

  it('when multiple TableRows are selected, selected state contains items', () => {
    wrapper.find(TableRow).forEach(n => n.simulate('click'));
    const ids = props.data.map(({ id }) => id);
    expect(wrapper.state('selected')).toEqual(ids);
  });

  it('when multiple TableRows are unselected, selected state does not contain items', () => {
    wrapper.find(TableRow).forEach(n => n.simulate('click'));
    wrapper.find(TableRow).forEach(n => n.simulate('click'));
    expect(wrapper.state('selected')).toEqual([]);
  });

  it('when TablePagination prop onChangePage is clicked, page state is updated', () => {
    wrapper.find(TablePagination).simulate('click');
    expect(wrapper.state('page')).toBe(0);
  });
});
