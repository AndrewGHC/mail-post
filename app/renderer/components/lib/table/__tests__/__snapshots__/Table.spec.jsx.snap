// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table matches snapshot 1`] = `
<WithStyles(Paper)>
  <WithStyles(EnhancedTableToolbar)
    numSelected={0}
    onClickDelete={[Function]}
  />
  <div>
    <WithStyles(Table)>
      <EnhancedTableHead
        numSelected={0}
        onRequestSort={[Function]}
        onSelectAllClick={[Function]}
        order="desc"
        orderBy="createdAt"
        rowCount={2}
      />
      <WithStyles(TableBody)>
        <WithStyles(TableRow)
          aria-checked={false}
          hover={true}
          key="1"
          onClick={[Function]}
          role="checkbox"
          selected={false}
          tabIndex={-1}
        >
          <WithStyles(TableCell)
            padding="checkbox"
          >
            <WithStyles(Checkbox)
              checked={false}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            padding="none"
          >
            name
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            10
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            March 26, 2015 6:30 PM
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
        <WithStyles(TableRow)
          aria-checked={false}
          hover={true}
          key="2"
          onClick={[Function]}
          role="checkbox"
          selected={false}
          tabIndex={-1}
        >
          <WithStyles(TableCell)
            padding="checkbox"
          >
            <WithStyles(Checkbox)
              checked={false}
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            padding="none"
          >
            name
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            10
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            numeric={true}
          >
            March 26, 2015 6:30 PM
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableBody)>
    </WithStyles(Table)>
  </div>
  <WithStyles(TablePagination)
    backIconButtonProps={
      Object {
        "aria-label": "Previous Page",
      }
    }
    component="div"
    count={2}
    nextIconButtonProps={
      Object {
        "aria-label": "Next Page",
      }
    }
    onChangePage={[Function]}
    onChangeRowsPerPage={[Function]}
    page={0}
    rowsPerPage={5}
  />
</WithStyles(Paper)>
`;
